src/targets/maven.ts

```diff
@@ -0,0 +1,389 @@
+    this.logger.debug(
+      `Checking if Gradle CLI is available on ${this.mavenConfig.gradleCliPath}...`
+    );
+    checkExecutableIsPresent(this.mavenConfig.gradleCliPath);
+    this.logger.debug(`Checking if GPG is available in the path...`);
+    checkExecutableIsPresent('gpg');
+  }
+
+  /**
+   * Parses the required target's configuration variables' types from strings to
+   * RegExp.
+   *
+   * To make TS happy, they must be `RegExp` objects. It's not valid to write
+   * `/myRegex/`
+   */
+  private makeRegexpFromConfiguration(): void {
+    this.config.androidDistDirPattern = new RegExp(
+      this.config.androidDistDirPattern
+    );
+    this.config.androidFileSearchPattern = new RegExp(
+      this.config.androidFileSearchPattern
+    );
+  }
+
+  /**
+   * Publishes current Java and Android distributions.
+   * @param version New version to be released.
+   * @param revision Git commit SHA to be published.
+   */
+  public async publish(_version: string, revison: string): Promise<void> {
+    await this.createUserGradlePropsFile();
+    await this.upload(revison);
+    await this.closeAndRelease();
+  }
+
+  /**
+   * Creates the required user's `gradle.properties` file.
+   *
+   * If there's an existing one, it's overwritten.
+   * TODO: control when it's overwritten with an option.
+   */
+  private async createUserGradlePropsFile(): Promise<void> {
+    await fsPromises.writeFile(
+      join(this.getGradleHomeDir(), GRADLE_PROPERTIES_FILENAME),
+      // Using `` instead of string concatenation makes all the lines but the
+      // first one to be indented to the right. To avoid that, these lines
+      // shouldn't have that much space at the beginning, something the linter
+      // doesn't agree with (and the code would be harder to read).
+      `mavenCentralUsername=${this.mavenConfig.MAVEN_CENTRAL_USERNAME}\n` +
+        `mavenCentralPassword=${this.mavenConfig.MAVEN_CENTRAL_PASSWORD}`
+    );
+  }
+
+  /**
+   * Retrieves the Gradle Home path.
+   *
+   * See https://docs.gradle.org/current/userguide/build_environment.html#sec:gradle_environment_variables
+   *
+   * @returns the gradle home path.
+   */
+  public getGradleHomeDir(): string {
+    if (process.env.GRADLE_USER_HOME) {
+      return process.env.GRADLE_USER_HOME;
+    }
+
+    return join(homedir(), '.gradle');
+  }
+
+  /**
+   * Uploads the artifacts with the required files. This is a required step
+   * to make a release, but this doesn't perform any releases; after upload,
+   * the flow must finish with `closeAndRelease`.
+   */
+  public async upload(revision: string): Promise<void> {
+    this.logger.debug('Fetching artifact list...');
+    const artifacts = await this.getArtifactsForRevision(revision);
+
+    await withTempDir(
+      async dir => {
+        await Promise.all(
+          artifacts.map(
+            async artifact => await this.uploadArtifact(artifact, dir)
+          )
+        );
+      },
```

Comment:
You are overusing `await`. You can just pass promises around (like pointers in C) which would save cycles. Every `await` means another cycle on the event loop to wait until the promise is resolved only to return yet another result with an empty value in this case
```suggestion
      dir => Promise.all(artifacts.map(artifact => this.uploadArtifact(artifact, dir))),
```